### PR TITLE
Only date of oath can be changed now (Fix #1843)

### LINKED ISSUE #1843: [API/Персональний профіль] user (except Registered user) has opportunity to edit his 'Дата вступу/присяги/завершення' in 'Дійсне членство' tab by using Postman
Environment: Windows 10 Home 64 bit 1903, Google Chrome Version 94.0.4606.61 (64 bit)
Reproducible: always
Priority: medium
Severity: medium
Build found: 26f3509

**Precondition**
Go to https://eplast.westeurope.cloudapp.azure.com

**Steps to reproduce**

1. Log into the system as User (except Registered user)
2. Press 'F12' button and click on 'Network' tab
3. Click on 'Дійсне членство' tab
4. Click on 'Змінити' button
5. Click on 'Змінити' button on the pop-up-menu
6. Type in 'Filter' textbox Method:POST
7. Click on 'dates' request and copy Request URL link, then put it in Postman
8. Copy User's token and put it in Postman on 'Authorization' tab (choose Bearer token in type)
9. Copy Request Payload on the bottom of Dev Tools and put it in Postman on 'body' tab
10. Edit User's 'Дата вступу/присяги/завершення' by using Postman

**Actual result**
User (except Registered user) has opportunity to edit his 'Дата вступу/присяги/завершення' by using Postman
change date act

**Expected result**
User does not have opportunity to edit 'Дата вступу/присяги/завершення' by using Postman. 'Дата вступу' should be earlier than 'Дата присяги/завершення'. 'Дата присяги' should be earlier than 'Дата завершення'
change date exp

### PR DESCRIPTION
**Pull request by petromonka**
We can change only the oath date now.
If user became former member when API will automatically set end date as  current date.
If user change role from former to registered when end date will be reset.

dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by dashapenkova — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs [hunk 1]
```diff
@@ -98,6 +98,13 @@ public async Task ChangeAsync(string userId)
                 await _sectorAdministrationService.RemoveAdministratorAsync(sectorAdmin.Id);
             }
 
+            var membershipDates = await _repoWrapper.UserMembershipDates.GetFirstOrDefaultAsync(m => m.UserId == userId);
+            if (membershipDates != null)
+            {
+                membershipDates.DateEnd = DateTime.Now;
+                _repoWrapper.UserMembershipDates.Update(membershipDates);
```
Comment (OleksandrBilan) [inline EPlast/EPlast.BLL/Services/Admin/AdminService.cs:105]: don't you need to call 'await _repoWrapper.SaveAsync();' ?

File: EPlast/EPlast.Tests/Controllers/ActiveMembershipControllerTests.cs [hunk 2]
```diff
@@ -289,49 +289,72 @@ public async Task GetUserDates_InValid_ThrowException()
...
+        public async Task ChangeUserOathDate_NoAccess_Return403Forbidden()
+        {
+            //Arrange
+            bool successfulChangedDates = false;
+            _userManager.Setup(x => x.GetUserAsync(It.IsAny<ClaimsPrincipal>())).ReturnsAsync(new User() { Id = _userId });
+            _userService.Setup(x => x.GetUserAsync(It.IsAny<string>())).ReturnsAsync(_user);
+            _userManager.Setup(x => x.GetRolesAsync(It.IsAny<User>())).ReturnsAsync(new List<string>() { Roles.PlastMember });
+            _userDatesService.Setup(cs => cs.ChangeUserOathDateAsync(It.IsAny<UserOathDateDTO>()))
+                             .ReturnsAsync(successfulChangedDates);
+            ActiveMembershipController activeMembershipController = _activeMembershipController;
```
Comment (OleksandrBilan) [inline EPlast/EPlast.Tests/Controllers/ActiveMembershipControllerTests.cs:345]: why do you need this local variable, can't you just use '_activeMembershipController'?

File: EPlast/EPlast.WebApi/Controllers/ActiveMembershipController.cs [hunk 3]
```diff
@@ -138,17 +138,23 @@ public async Task<IActionResult> GetUserDates(string userId)
...
-            if (!await HasAccessAsync(userMembershipDatesDTO.UserId))
+            var focusUser = await _userManager.FindByIdAsync(userOathDateDTO.UserId);
+            var roles = await _userManager.GetRolesAsync(focusUser);
+            //If user is registered or former member, then we can not change oath date!
+            if(roles.Contains(Roles.RegisteredUser) || roles.Count == 0)
             {
                 return StatusCode(StatusCodes.Status403Forbidden);
             }
-            if (await _userDatesService.ChangeUserMembershipDatesAsync(userMembershipDatesDTO))
+            if (!await HasAccessAsync(userOathDateDTO.UserId))
```
Comment (OleksandrBilan) [inline EPlast/EPlast.WebApi/Controllers/ActiveMembershipController.cs:150]: you can add this check to the previous if statement

File: EPlast/EPlast.BLL/Services/Admin/AdminService.cs (hunk 1 above)
Comment (petromonka) [reply EPlast/EPlast.BLL/Services/Admin/AdminService.cs:105]: Totally forgot about Save Async. Thank you for the right remark 👍

File: EPlast/EPlast.Tests/Controllers/ActiveMembershipControllerTests.cs (hunk 2 above)
Comment (petromonka) [reply EPlast/EPlast.Tests/Controllers/ActiveMembershipControllerTests.cs]: Removed extra variable.

File: EPlast/EPlast.WebApi/Controllers/ActiveMembershipController.cs (hunk 3 above)
Comment (petromonka) [reply EPlast/EPlast.WebApi/Controllers/ActiveMembershipController.cs]: Great advice. After your review, my impostor syndrome grew up ;(

**Review by VolodymyrKostiv — APPROVED** [review]

**Review by OleksandrBilan — APPROVED** [review]

**Pull request merged**